File: sdk/elixir/test/ci_test.exs
defmodule Bsdkrun.CITest do
  use ExUnit.Case, async: true

  alias Bsdkrun.CI

  # The YAML the builder emits is consumed by tangled's own workflow parser
  # (inside `bsdkrun ci`), so these tests pin the emitted shape — a change
  # here is a change to what spindle would receive.

  test "renders the full workflow shape" do
    y =
      CI.workflow("test")
      |> CI.on_push("main")
      |> CI.on_pull_request(["main", "develop"])
      |> CI.deps(["elixir", "erlang"])
      |> CI.deps_from("github:nix-community/fenix/abc123", ["stable.default"])
      |> CI.env("MIX_ENV", "test")
      |> CI.clone_depth(100)
      |> CI.step("deps", "mix deps.get")
      |> CI.step("test", "mix test", %{"WARNINGS_AS_ERRORS" => "1"})
      |> CI.yaml()

    assert y =~ "  - event: [\"push\"]\n    branch: \"main\""
    assert y =~ "branch: [\"main\", \"develop\"]"
    assert y =~ "engine: nixery"
    assert y =~ "\"nixpkgs\":\n    - \"elixir\"\n    - \"erlang\""
    assert y =~ "\"github:nix-community/fenix/abc123\":"
    assert y =~ "MIX_ENV: \"test\""
    assert y =~ "depth: 100"
    assert y =~ "- name: \"deps\"\n    command: |\n      mix deps.get"
    assert y =~ "environment:\n      WARNINGS_AS_ERRORS: \"1\""
  end

  test "block-unsafe commands fall back to a JSON string" do
    # Trailing spaces do not survive a literal block scalar; the emitter must
    # switch representation rather than silently altering the command.
    y = CI.workflow("edge") |> CI.step("tricky", "echo 'a'  \necho b") |> CI.yaml()
    assert y =~ "command: \"echo 'a'  \\necho b\""
  end

  test "file names get the yml suffix" do
    assert CI.workflow("build") |> CI.file_name() == "build.yml"
    assert CI.workflow("build.yaml") |> CI.file_name() == "build.yaml"
  end
end

File: sdk/typescript/test/ci.test.ts
import { describe, expect, it } from "bun:test";
import { workflow } from "../src/ci.js";

// The YAML this builder emits is consumed by tangled's own workflow parser
// (inside `bsdkrun ci`), so these tests pin the emitted shape — a change here
// is a change to what spindle would receive.
describe("ci workflow builder", () => {
  it("renders the full workflow shape", () => {
    const y = workflow("test")
      .onPush("main")
      .onPullRequest("main", "develop")
      .deps("nodejs", "pnpm")
      .depsFrom("github:nix-community/fenix/abc123", "stable.defaultToolchain")
      .env("CI_FROM", "sdk")
      .cloneDepth(100)
      .step("install", "pnpm install")
      .step("test", "pnpm test", { NODE_ENV: "test" })
      .yaml();

    expect(y).toContain('  - event: ["push"]\n    branch: "main"');
    expect(y).toContain('branch: ["main", "develop"]');
    expect(y).toContain("engine: nixery");
    expect(y).toContain('"nixpkgs":\n    - "nodejs"\n    - "pnpm"');
    expect(y).toContain('"github:nix-community/fenix/abc123":');
    expect(y).toContain('CI_FROM: "sdk"');
    expect(y).toContain("depth: 100");
    expect(y).toContain('- name: "install"\n    command: |\n      pnpm install');
    expect(y).toContain('environment:\n      NODE_ENV: "test"');
  });

  it("falls back to a JSON string for block-unsafe commands", () => {
    // Trailing spaces do not survive a literal block scalar; the emitter
    // must switch representation rather than silently altering the command.
    const y = workflow("edge").step("tricky", "echo 'a'  \necho b").yaml();
    expect(y).toContain('command: "echo \'a\'  \\necho b"');
  });

  it("derives the file name", () => {
    expect(workflow("build").fileName()).toBe("build.yml");
    expect(workflow("build.yaml").fileName()).toBe("build.yaml");
  });
});
